juju/juju
# [JUJU-1263] Updated storage params code start

# Introduction

What kicked off this change was the Juju was know longer able to
bootstrap to Azure clusters because the predefined notion Juju had about
what storage to use by default was know longer correct because Azure had
changed the naming of their provisioner's.

Upon closer inspection of the logic the code had a number of short
comings and bugs that did not allow it operate in the manner it was
originally intended to do.

This change is big but simplifies the logic used for finding storage in
Kubernetes now. Specifically cluster storage is now laid out
in provider preferences where the provider can find the best matching
storage class based on it's list. This allows the underlying cloud code
to be more agile and tolerate change from the underlying cloud and also
enshrines the logic used previously into a more robust readable
solution.

## Checklist


## QA steps

This is a pretty substantial change and manual testing is fine but we really want to check the unit tests here because they setup a fake Kubernetes with the different pre states we are expecting to encounter. What we want to do is check that these pre-state's are matched to post states from GetClusterMetadata that we expect to see.

If a manual test is required then you may bootstrap against a fresh Azure Kubernetes cluster to check that bootstrapping is successful as this was the original problem that inspired this work.

## Documentation changes

Currently investigating. Will update PR once I have looked.

## Bug reference

Original bug as been fixed on the 2.9 branch. This was work to make the overall fix better but not introduce to much churn on the 2.9 stream.
